### PR TITLE
[BEAM-1139] Add no-arg constructor for UnboundedReadFromBoundedSource

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This allows Kryo to work with the type, currently required by
the Apex runner.

### CODE REVIEW

**Comment by tgroh** [comment]
My current exception 
```
java.lang.InstantiationError: org.apache.beam.sdk.io.UnboundedSource

	at org.apache.beam.sdk.io.UnboundedSourceConstructorAccess.newInstance(Unknown Source)
	at com.esotericsoftware.kryo.Kryo$1.newInstance(Kryo.java:1015)
	at com.esotericsoftware.kryo.Kryo.newInstance(Kryo.java:1065)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.create(FieldSerializer.java:228)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:217)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:626)
	at org.apache.beam.runners.core.UnboundedReadFromBoundedSourceTest.testAdapterKryoSerialization(UnboundedReadFromBoundedSourceTest.java:143)
```

**Comment by tgroh** [comment]
I get that with a public or private no-arg constructor, with a final or non-final class, and with a final or non-final `boundedSource` field.

**Comment by tgroh** [comment]
R: @tweise @lukecwik 

The serialization test passes in the DirectRunner Module locally.

**Comment by kennknowles** [comment]
Confirmed the fix for just the error in question manually, since Jenkins is severely backlogged.

```
mvn -DskipTests clean install -P apex-runner -pl examples/java -am

mvn --errors failsafe:integration-test@apex-runner-integration-tests \
    -pl examples/java \
    -P apex-runner \
    -P jenkins-precommit \
    -D test=WordCountIT \
    -D beamTestPipelineOptions='[
      "--runner=org.apache.beam.runners.apex.TestApexRunner",
      "--project=...",
      "--tempRoot=..."
    ]'
```

File: runners/direct-java/pom.xml [hunk 1]
```diff
@@ -262,7 +262,15 @@
       </exclusions>
       <scope>test</scope>
     </dependency>
-    
+
+    <!-- Required for UnboundedReadFromBoundedSourceTest -->
+    <dependency>
+      <groupId>com.esotericsoftware.kryo</groupId>
+      <artifactId>kryo</artifactId>
+      <version>2.21</version>
```
Comment (dhalperi) [inline runners/direct-java/pom.xml]: please do not put versions in submodules -- top module only.

Comment (dhalperi) [inline runners/direct-java/pom.xml]: why is this needed at all -- isn't it picked up from the fact that this depends on runner-core?

**Comment by davorbonaci** [comment]
Can we investigate alternatives that don't involve kryo proliferation? I'd ask to please hold the PR in the meanwhile.

**Comment by kennknowles** [comment]
@davorbonaci this is precommit breakage

**Comment by lukecwik** [comment]
@davorbonaci There is currently no hard specification as to how objects need to be serialized. Kryo proliferation comes from runners relying on Kryo as their defacto serialization mechanism.

**Comment by dhalperi** [comment]
But Spark, for example, serializes sources and coders using java serialization as is expected today by Beam.

**Comment by tweise** [comment]
I'm going to look at it tonight. I believe Java serialization can be used for this reference in the runner also.

**Comment by kennknowles** [comment]
Thanks @tweise! Agree that we should hold off on a quick fix, then, if you think you can just avoid the whole issue. It will save us debating the relationship of Kryo and Beam :-)

**Comment by davorbonaci** [comment]
LGTM. Good for self-merge.

**Comment by tweise** [comment]
We can potentially improve it later to. Generally, if the class is Java serializable, then use

`@Bind(JavaSerializer.class)`

in the Apex operator code.

**Comment by tweise** [comment]
I'm looking at it, not sure what the issue is, the field is marked for Java serialization in master:

```
Caused by: com.esotericsoftware.kryo.KryoException: Class cannot be created (missing no-arg constructor): org.apache.beam.sdk.io.CompressedSource
Serialization trace:
boundedSource (org.apache.beam.runners.core.UnboundedReadFromBoundedSource$BoundedToUnboundedSourceAdapter)
source (org.apache.beam.runners.apex.translation.operators.ApexReadUnboundedInputOperator)
	at com.esotericsoftware.kryo.Kryo.newInstantiator(Kryo.java:1050)
```

```
  @Bind(JavaSerializer.class)
  private final UnboundedSource<OutputT, CheckpointMarkT> source;
```

There should be no Kryo exception.

File: runners/core-java/pom.xml [hunk 2]
```diff
@@ -152,6 +152,13 @@
 
     <!-- test dependencies -->
 
+    <dependency>
+      <groupId>com.esotericsoftware.kryo</groupId>
+      <artifactId>kryo</artifactId>
+      <version>2.21</version>
```
Comment (tweise) [inline runners/core-java/pom.xml]: There shoudn't be any Kryo dependency in runners core. If at all, then this test should be in the Apex runner.

**Comment by tweise** [comment]
I notice that the @Bind annotation that controls the serializer is only present in Kryo 2.24 (used by Apex), not in 2.21. Is there a dependency leak in the examples module?

**Comment by dhalperi** [comment]
Sure looks like it. This may indicate a flaw in our testing strategy. We
have K profiles for the K runners we have, and they work fine with proper
transitive dependency resolution when used individually. Precommit,
however, we simply turn on all the profiles and hope that enabling them all
concurrently just works. (Which it mostly, sorta, has.)
